### DISCUSSION TITLE
Recommended way to set prototype of custom JavaScript type?

### DISCUSSION THREAD

**Discussion by EnCey** [sebastienros/jint#1364] [Q&A] [answered]
I've followed the example outlined in the [Wiki](https://github.com/sebastienros/jint/wiki#custom-javascript-buildin-types-for-jint-v3) about [`UuidInstance`](https://github.com/sebastienros/jint/blob/11b0173d6ad78226bf22f56d615b94eb64b9deb0/Jint.Tests/Runtime/Domain/UuidInstance.cs#L6).

It shadows the internal `_prototype` field:

```
protected override ObjectInstance GetPrototypeOf() => _prototype;
internal ObjectInstance _prototype;
```

However, I noticed that `ObjectInstance` uses the internal `_prototype` field in other methods as well, like `SetPrototypeOf`. 
Won't I get in hell's kitchen by only partially introducing my own `_prototype` field?

I'm wondering why not just to call `SetPrototypeOf` on the `UuidInstance` in  `UuidConstructor`, i.e. instead of:
https://github.com/sebastienros/jint/blob/11b0173d6ad78226bf22f56d615b94eb64b9deb0/Jint.Tests/Runtime/Domain/UuidConstructor.cs#L66

something like this:

```
public UuidInstance Construct(JsUuid uuid) {
  var instance = new UuidInstance(Engine) { PrimitiveValue = uuid };
  instance.SetPrototypeOf(PrototypeObject);
  return instance;
}
```

**update:** it just occurs to me, if not for the `PrimitiveValue` property on `UuidInstance`, do I even need to create a custom implementation of `ObjectInstance` ? Wouldn't a Prototype and Consturctor suffice?

**Comment by lahma** [comment, marked as answer]
Here's an alternative approach jos JS UUID kind of integration: https://github.com/sebastienros/jint/pull/1183 . Options has wrap object handler callback that you could also use to set prototypes.

**Reply by EnCey** [reply]
Thanks for the tips. I'm trying to create an object with methods that return promises, which internally call async .NET methods:  `foo.load(...).then(result => ...)`
Using `TypeDescriptor` / `TypeReference` to build a mapping to async .NET won't work out of the box, I need to inject some custom code that takes care of the Task/Promise handling. 

Do you see a problem if I use `SetPrototypeOf()` on my custom `ObjectInstances` ?

**Reply by EnCey** [reply]
I only have a very rudimentary understanding of the API as of yet, but could the `ObjectInstance` ctor perhaps be extended to allow passing a specific prototype?

```diff
        internal ObjectInstance(
            Engine engine,
            ObjectClass objectClass = ObjectClass.Object,
            InternalTypes type = InternalTypes.Object,
+           ObjectInstance? prototype = null)
            : base(type)
        {
            _engine = engine;
            _class = objectClass;
            // if engine is ready, we can take default prototype for object
-           _prototype = engine.Realm.Intrinsics?.Object?.PrototypeObject;
+           _prototype = prototype ?? engine.Realm.Intrinsics?.Object?.PrototypeObject;
            Extensible = true;
        }
```

**Reply by lahma** [reply]
`SetPrototypeOf` is the "standard" way of doing things as [it's also exposed on JS side](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf). So I would recommend using it unless there's some blocker I'm not aware of. C# interop can have it rough edges, especially as you are working with the async world here.

**Reply by EnCey** [reply]
👍 I'll go with `SetPrototypeOf` then
